### ISSUE TITLE
Problems sending raw data with netmap

### ISSUE THREAD

**Issue by prophetHM** [luigirizzo/netmap#467]
Hi,

I'm currently trying to send a few raw packets using netmap within my own application. Previously, I was using the `pkt-gen` app which was working fine. Therefore, I have extracted the essential code from the pkt-gen tool, but the raw packets are not sent. In fact, `nm_tx_pending` always returns 1. I guess I am making a serious mistake in my code, but I'm not able to find it. Hope you guys can help me further.

This is the code:

```C
int i;
uint8_t buffer[ETH_FRAME_LEN];

uint8_t*  etherhead = buffer;
uint8_t* data = buffer + sizeof(struct ethhdr);
 
struct ethhdr *eh = (struct ethhdr *) etherhead;

/* Ethernet Header Construction */
for (i = 0; i < ETHER_ADDR_LEN; ++i)
{
   eh->h_dest[i] 	= dest_mac[i];
   eh->h_source[i] = src_mac[i];
}
eh->h_proto = htons(/* some raw protocol */)

int size = sizeof(struct ethhdr) + pkt_len;

struct nm_desc *nmd = nm_open("netmap:eth0", NULL, 0, NULL);
if(nmd == NULL)
	return -1;
	
struct netmap_if *nifp = nmd->nifp;
struct netmap_ring *txring = NETMAP_TXRING(nifp, nmd->first_tx_ring);
if(txring == NULL)
	return -2;

int cur = txring->head;
for(i=0; i<nrPackets; i++)
{
    struct netmap_slot *slot = &txring->slot[cur];
    char *p = NETMAP_BUF(txring, slot->buf_idx);
	
    slot->flags = 0;
    nm_pkt_copy(buffer, p, size);
	
    if (i == count - 1) {
        slot->flags |= NS_REPORT;
    }

    cur = nm_ring_next(txring, cur);
}

txring->head = txring->cur = cur;

ioctl(nmd->fd, NIOCTXSYNC, NULL);

while(nm_tx_pending(txring)) { //XXX: always returns 1
    ioctl(nmd->fd, NIOCTXSYNC, NULL);
    fprintf(stderr, "Still pending...\n");
    usleep(1);
}

nm_close(nmd);
```

Thanks in advance!

**Comment by giuseppelettieri** [comment]
First check that your program works with software only ports, e.g. send on ```vale0:0``` and read from ```vale0:1``` (you can read with ```pkt-gen```, of course).

It the above works, the problem with real NICs is probably that you are starting to transmit as soon as you ```nm_open()``` the port. Unfortunately, in order to switch to netmap mode, we have to turn the interface down and then up. This takes a few seconds, during which the NIC is ignoring your requests. Putting a ```sleep(2)``` (or more) after ```nm_open()``` should fix your problem.

**Comment by prophetHM** [comment]
Indeed, waiting for a few seconds fixed my problem! My application now works as expected, thank you!

Just as a side question, could it be realized that the NIC is switched to netmap mode during system boot? I guess this would limit the NIC to netmap-only during uptime.

**Comment by giuseppelettieri** [comment]
Not in the current implementation. You need to ```nm_open()``` a port to switch it to netmap mode. Of course, you can write a daemon that does just that and never terminates. This will prevent others to re-open the same port in "exclusive mode" (flag ```/x``` in the name passed to ```nm_open()```), but normal open will be allowed.

**Comment by vmaffione** [comment]
Instead from starting from pkt-gen you should look at the netmap tutorial and its example programs
Https://github.com/netmap-unipi/netmap-tutorial
